File: Chapter07/Java/RESTJSON/app/src/main/java/net/zenconsult/android/RESTJSONActivity.java
package net.zenconsult.android;

import android.app.ListActivity;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.widget.ArrayAdapter;
import android.widget.TextView;
import android.widget.Toast;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

public class RESTJSONActivity extends ListActivity {

    private static final String TAG = "RESTJSONActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        getListView().setTextFilterEnabled(true);
        getListView().setOnItemClickListener((parent, view, position, id) -> {
            Toast.makeText(this, ((TextView)view).getText(), Toast.LENGTH_SHORT).show();
        });

        new AsyncTask<Void, Void, String[]>() {

            @Override
            protected String[] doInBackground(Void... voids) {
                try {
                    URL url = new URL("http://www.amazingrace.cn/apress/members");
                    HttpURLConnection connection = (HttpURLConnection) url.openConnection();

                    connection.setRequestMethod("GET");
                    connection.setDoOutput(false);
                    connection.setDoInput(true);
                    connection.setUseCaches(false);

                    connection.connect();

                    Log.d(TAG, "doInBackground=>responseCode: " + connection.getResponseCode());
                    if (connection.getResponseCode() == HttpURLConnection.HTTP_OK) {
                        StringBuilder sb = new StringBuilder();
                        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
                        String line = null;
                        while ((line = reader.readLine()) != null) {
                            sb.append(line);
                        }
                        reader.close();

                        Log.d(TAG, "doInBackground=>response: " + sb.toString());

                        JSONObject users = new JSONObject(sb.toString()).getJSONObject("users");
                        JSONArray jArray = users.getJSONArray("user");
                        String[] names = new String[jArray.length()];
                        for (int i = 0; i < jArray.length(); i++) {
                            JSONObject jsonObject = jArray.getJSONObject(i);
                            names[i] = jsonObject.getString("name");
                        }
                        return names;
                    }
                } catch (Exception e) {
                    Log.e(TAG, "doInBackground=>error: ", e);
                }
                return null;
            }

            @Override
            protected void onPostExecute(String[] result) {
                super.onPostExecute(result);
                if (result == null) {
                    Log.e(TAG, "onPostExecute=>result is null.");
                    return;
                }
                setListAdapter(new ArrayAdapter<String>(RESTJSONActivity.this, R.layout.list_item, result));
            }

        }.execute();
    }
}

File: Chapter07/Kotlin/RESTJSON/app/src/main/java/net/zenconsult/android/RESTJSONActivity.kt
package net.zenconsult.android

import android.app.ListActivity
import android.os.AsyncTask
import android.os.Bundle
import android.util.Log
import android.widget.ArrayAdapter
import android.widget.TextView
import android.widget.Toast
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.lang.Exception
import java.lang.StringBuilder
import java.net.HttpURLConnection
import java.net.URL

class RESTJSONActivity : ListActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        listView.isTextFilterEnabled = true
        listView.setOnItemClickListener { _, view, _, _ ->
            Toast.makeText(this@RESTJSONActivity, (view as TextView).text, Toast.LENGTH_SHORT)
                .show()
        }

        FetchTask(this).execute()
    }

    private class FetchTask(
        val activity: RESTJSONActivity
    ): AsyncTask<Any?, Any?, ArrayList<String>>() {

        override fun doInBackground(vararg params: Any?): ArrayList<String>? {
            try {
                val url = URL("http://www.amazingrace.cn/apress/members")
                val connection = url.openConnection() as HttpURLConnection

                connection.apply {
                    requestMethod = "GET"
                    doOutput = false
                    doInput = true
                    useCaches = false
                }

                connection.connect()

                Log.d(TAG, "doInBackground=>responseCode: ${connection.responseCode}")
                if (connection.responseCode == HttpURLConnection.HTTP_OK) {
                    val sb = StringBuilder()
                    val reader = BufferedReader(InputStreamReader(connection.inputStream))
                    var line = reader.readLine()
                    while (line != null) {
                        sb.append(line)
                        line = reader.readLine()
                    }
                    reader.close()

                    Log.d(TAG, "doInBackground=>response: $sb")

                    val users = JSONObject(sb.toString()).getJSONObject("users")
                    val jArray = users.getJSONArray("user")
                    val names = ArrayList<String>()
                    var i = 0
                    while (i < jArray.length()) {
                        val jsonObject = jArray.getJSONObject(i)
                        names.add(jsonObject.getString("name"))
                        i++
                    }
                    return names
                }
            } catch (e: Exception) {
                Log.e(TAG, "doInBackground=>error: ", e)
            }
            return null
        }

        override fun onPostExecute(result: ArrayList<String>?) {
            super.onPostExecute(result)
            if (result == null) {
                Log.e(TAG, "onPostExecute=>result is null.")
                return;
            }
            result?.let {
                activity.listAdapter = ArrayAdapter<String>(activity, R.layout.list_item, it)
            }
        }

    }

    companion object {
        const val TAG = "RESTJSONActivity"
    }
}